### ISSUE TITLE
QoS Misclassification

### ISSUE THREAD

**Issue by farnwomt** [Jackysi/advancedtomato#238]
I have noticed an apparent fault with the QoS system.

When doing a large download I can go to the "Quality of Service->View Details" and I can see that the connection is classified as a FileXfer.

However when I go to "Quality of Service->View Graphs" I can see on the graph that the connection is classified as P2PBulk.

Furthermore it is quite clear that the bandwidth is being limited a though it is a P2PBulk connection and limited to 30% unlike a FileXfer which has a limit of 100%.

I am using 1.28.0000 MIPSR2-3.3-138 K26 USB Mega-VPN

**Comment by Jackysi** [comment]
Sounds more like firmware issue than GUI. I will check where these things are defined, but I'm pretty sure its not in the GUI.

**Comment by farnwomt** [comment]
Having looked a little more closely I think that whilst outgoing traffic is split up all my incoming traffic is being classified as P2PBulk.  Not sure why, but I am happy to report it elsewhere if you can point me in the right direction.

Thanks,
Michael

**Comment by f00b4r0** [comment]
Hi,

I have the exact same problem (on v3.3-138 Mega-VPN), and I've identified the cause: in /etc/wan_qos, all IMQ IPV4 filters miss the "protocol ip" argument:

```
  # class id 90: rate 30kbit ceil 300kbit
  $TCA_IMQ parent 1:1 classid 1:90 htb rate 30kbit ceil 300kbit prio 9 quantum 1500
  $TQA_IMQ parent 1:90 handle 90: $Q
  $TFA_IMQ parent 1: prio 90 handle 9 fw flowid 1:90 
  $TFA_IMQ parent 1: prio 190 protocol ipv6 handle 9 fw flowid 1:90 
```

adding "protocol ip" in the first $TFA_IMQ line fixes the bug and makes inbound classifier work. (Running qos_wan without this change spits a lot of errors, with the change no errors are emitted)

I don't know if that's an upstream bug, but at least I hope this info will help quickly fix it.

**Comment by Jackysi** [comment]
https://bitbucket.org/pl_shibby/tomato-arm/issues?status=new&status=open!

**Comment by farnwomt** [comment]
I really appreciate your efforts in locating this issue.  Obviously QoS 
is one of the big reasons to use Tomato, so having this fixed makes a 
big difference.

**Comment by f00b4r0** [comment]
You're welcome. I have added extra information about this issue and a fix suggestion here: [bitbucket/tomato-arm:#74](https://bitbucket.org/pl_shibby/tomato-arm/issues/74/)

**Comment by farnwomt** [comment]
Unfortunately this issue with the upstream code hasn't been fixed yet and it really does mean that AdvancedTomato doesn't work well for QoS if you are using the latest code.

The hack to fix it is to add some code to Administration->Scripts->Firewall.

I tried to post the code to add here, but the formatting system of this website makes a mess of it, so please take a look at the link in the posting above.  This isn't the right solution, but it is a fix for the time being!
